java/ql/lib/semmle/code/java/os/OSCheck.qll

```diff
@@ -0,0 +1,158 @@
+private import semmle.code.java.environment.SystemProperty
+private import semmle.code.java.frameworks.apache.Lang
+private import semmle.code.java.dataflow.DataFlow
+
+/**
+ * A guard that checks if the current os is Windows.
+ * When True, the OS is Windows.
+ * When False, the OS is not Windows.
+ */
+abstract class IsWindowsGuard extends Guard { }
+
+/**
+ * A guard that checks if the current OS is any Windows.
+ * When True, the OS is Windows.
+ * When False, the OS *may* still be Windows.
+ */
+abstract class IsAnyWindowsGuard extends Guard { }
+
+/**
+ * A guard that checks if the current OS is unix or unix-like.
+ * When True, the OS is unix or unix-like.
+ * When False, the OS is not unix or unix-like.
+ */
+abstract class IsUnixGuard extends Guard { }
+
+/**
+ * A guard that checks if the current OS is unix or unix-like.
+ * When True, the OS is unix or unix-like.
+ * When False, the OS *may* still be unix or unix-like.
+ */
+abstract class IsAnyUnixGuard extends Guard { }
+
+/**
+ * Holds when the MethodAccess is a call to check the current OS using either the upper case `osUpperCase` or lower case `osLowerCase` string constants.
+ */
+bindingset[osString]
+private predicate isOsFromSystemProp(MethodAccess ma, string osString) {
+  exists(Expr systemGetPropertyExpr, Expr systemGetPropertyFlowsToExpr |
+    systemGetPropertyExpr = getSystemProperty("os.name")
+  |
+    DataFlow::localExprFlow(systemGetPropertyExpr, systemGetPropertyFlowsToExpr) and
+    ma.getAnArgument().(CompileTimeConstantExpr).getStringValue().toLowerCase().matches(osString) and // Call from System.getProperty to some partial match method
+    (
+      systemGetPropertyFlowsToExpr = ma.getQualifier()
+      or
+      exists(MethodAccess caseChangeMa |
+        caseChangeMa.getMethod() =
+          any(Method m |
+            m.getDeclaringType() instanceof TypeString and m.hasName(["toLowerCase", "toUpperCase"])
+          )
+      |
+        systemGetPropertyFlowsToExpr = caseChangeMa.getQualifier() and // Call from System.getProperty to case-switching method
+        DataFlow::localExprFlow(caseChangeMa, ma.getQualifier()) // Call from case-switching method to some partial match method
+      )
+    )
+  ) and
+  ma.getMethod() instanceof StringPartialMatchMethod
+}
+
+private class IsWindowsFromSystemProp extends IsWindowsGuard instanceof MethodAccess {
+  IsWindowsFromSystemProp() { isOsFromSystemProp(this, "window%") }
+}
+
+/**
+ * Holds when the Guard is an equality check between the system property with the name `propertyName`
+ * and the string or char constant `compareToLiteral`.
+ */
+private Guard isOsFromSystemPropertyEqualityCheck(string propertyName, string compareToLiteral) {
+  result
+      .isEquality(getSystemProperty(propertyName),
+        any(Literal literal |
+          (literal instanceof CharacterLiteral or literal instanceof StringLiteral) and
+          literal.getValue() = compareToLiteral
+        ), _)
```

Comment:
@smowton I could probably simplify this logic with the following diff to `CompileTimeConstantExpr`. Do I have your approval to make the following change?

```diff
diff --git a/java/ql/lib/semmle/code/java/Expr.qll b/java/ql/lib/semmle/code/java/Expr.qll
index 5a991d8814..d98cdbb6b0 100755
--- a/java/ql/lib/semmle/code/java/Expr.qll
+++ b/java/ql/lib/semmle/code/java/Expr.qll
@@ -168,6 +168,8 @@ class CompileTimeConstantExpr extends Expr {
   string getStringValue() {
     result = this.(StringLiteral).getValue()
     or
+    result = this.(CharacterLiteral).getValue()
+    or
     result =
       this.(AddExpr).getLeftOperand().(CompileTimeConstantExpr).getStringValue() +
         this.(AddExpr).getRightOperand().(CompileTimeConstantExpr).getStringValue()

```

Reply:
To make it easier, I broke this out into an independent PR: https://github.com/github/codeql/pull/8325